pkg/remoteimageasync/synchronizer.go

```diff
@@ -0,0 +1,119 @@
+package remoteimageasync
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/warm-metal/container-image-csi-driver/pkg/remoteimage"
+	"k8s.io/klog/v2"
+)
+
+// sessionChanDepth : 100 - must give lots of buffer to ensure no deadlock or dropped requests
+// completedChanDepth : 20 - must give some buffer to ensure no deadlock
+func StartAsyncPuller(ctx context.Context, sessionChanDepth, completedChanDepth int) AsyncPuller {
+	klog.Infof("%s.StartAsyncPuller(): starting async puller", prefix)
+	sessionChan := make(chan *PullSession, sessionChanDepth)
+	completedChan := make(chan string, completedChanDepth)
+	async := getSynchronizer(
+		ctx,
+		sessionChan,
+		completedChan,
+	)
+	async.RunCompletionsLoop()
+	RunPullerLoop(ctx, sessionChan, completedChan)
+	klog.Infof("%s.StartAsyncPuller(): async puller is operational", prefix)
+	return async
+}
+
+// channels are exposed for testing
+func getSynchronizer(
+	ctx context.Context,
+	sessionChan chan *PullSession,
+	completedChan chan string,
+) synchronizer {
+	if cap(sessionChan) < 50 {
+		klog.Fatalf("%s.getSynchronizer(): session channel must have capacity to buffer events, minimum of 50 is required", prefix)
+	}
+	if cap(completedChan) < 5 {
+		klog.Fatalf("%s.getSynchronizer(): completion channel must have capacity to buffer events, minimum of 5 is required", prefix)
+	}
+	return synchronizer{
+		sessionMap:      make(map[string]*PullSession),
+		mutex:           &sync.Mutex{},
+		sessions:        sessionChan,
+		completedEvents: completedChan,
+		ctx:             ctx,
+	}
+}
+
+func (s synchronizer) StartPull(image string, puller remoteimage.Puller, asyncPullTimeout time.Duration) (ses *PullSession, err error) {
```

Comment:
the `asyncPullTimeout` is passed into the `AsyncPuller.StartPull()` so that it can be varied by from image to image. this is in case we decide to allow that to be changed in the volume definition. this might not be desirable or possible so this could be removed as you say. this implementation could accept the same value every time or a different value per image so i exposed it that way. the other behavior was left to the next level up (node_server.go).

Reply:
you're correct, we could probably rely on extracting the image name from `puller.Image()`. i didn't take a dependency on that here.